proxy/src/http/id.rs

```diff
@@ -0,0 +1,266 @@
+//! Endpoints for Id.
+
+use serde::{Deserialize, Serialize};
+use warp::document::{self, ToDocumentedType};
+use warp::{path, Filter, Rejection, Reply};
+
+use crate::http;
+use crate::registry;
+
+/// `GET ids/<id>/status`
+pub fn get_status_filter<R: registry::Client>(
+    registry: http::Shared<R>,
+) -> impl Filter<Extract = impl Reply, Error = Rejection> + Clone
+where
+    R: registry::Client,
+{
+    path("ids")
+        .and(http::with_shared(registry))
+        .and(warp::get())
+        .and(document::param::<String>(
+            "id",
+            "The id whose status will be obtained",
+        ))
+        .and(path("status"))
+        .and(path::end())
+        .and(document::document(document::tag("Id")))
+        .and(document::document(document::description(
+            "Fetch the availability status of the given id",
+        )))
+        .and(document::document(
+            document::response(
+                200,
+                document::body(Status::document()).mime("application/json"),
+            )
+            .description("Successful retrieval"),
+        ))
+        .and_then(handler::get_status)
+}
+
+/// The status of an org or user id in the Registry.
+/// Only used for implementing `ToDocumentedType`.
+#[derive(Clone, Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub struct Status;
+
+impl ToDocumentedType for Status {
+    fn document() -> document::DocumentedType {
+        document::enum_string(vec!["available".into(), "taken".into(), "retired".into()])
+            .description("Variants for possible id statuses.")
+            .example("available")
+    }
+}
+
+/// Org handlers for conversion between core domain and http request fullfilment.
+#[allow(clippy::unwrap_used, clippy::indexing_slicing, clippy::panic, warnings)]
+mod handler {
+    use std::convert::TryFrom;
+    use warp::{reply, Rejection, Reply};
+
+    use crate::error::Error;
+    use crate::http;
+    use crate::registry;
+
+    /// Get the status for the given `id`.
+    pub async fn get_status<R: registry::Client>(
+        registry: http::Shared<R>,
+        id_string: String,
```

Comment:
Pretty sure you could do `registry::Id` here because of its `Deserialise` instance

Reply:
Good point, thanks for keeping on bringing that up.